File: drivers/keri-core/m63/hybrid_inception.py
"""Build M63 KERI-conformant hybrid icp — keripy reference (keri==1.1.17)."""

from __future__ import annotations

import base64
from dataclasses import dataclass
from typing import Any

from keri.core import coring, serdering
from keri.core.coring import Diger, Ilks, MtrDex, Serials, Verfer, versify
from keri.kering import Version

from .cesr import (
    CESR_MLDSA65_VERKEY,
    CESR_MLKEM768_ENCAP,
    CESR_X25519_PUBKEY,
    CIPHER_SUITE_IA_HYBRID_1,
    MLDSA65_VERKEY_BYTES,
    MLKEM768_ENCAP_BYTES,
    X25519_PUBKEY_BYTES,
    encode_large_fixed,
)


@dataclass(frozen=True)
class HybridKeyMaterial:
    """Raw key bytes for a hybrid inception (synthetic or caller-supplied)."""

    ed25519_signing_raw: bytes
    mldsa65_signing_raw: bytes
    x25519_agreement_raw: bytes
    mlkem768_encap_raw: bytes
    next_ed25519_signing_raw: bytes
    next_mldsa65_signing_raw: bytes


def synthetic_hybrid_key_material(seed: int = 0) -> HybridKeyMaterial:
    """Deterministic synthetic key material for harness vectors (not production keys)."""

    def fill(n: int, tag: int) -> bytes:
        return bytes([(seed + tag + i) % 256 for i in range(n)])

    return HybridKeyMaterial(
        ed25519_signing_raw=fill(32, 0x01),
        mldsa65_signing_raw=fill(MLDSA65_VERKEY_BYTES, 0x02),
        x25519_agreement_raw=fill(X25519_PUBKEY_BYTES, 0x03),
        mlkem768_encap_raw=fill(MLKEM768_ENCAP_BYTES, 0x04),
        next_ed25519_signing_raw=fill(32, 0x11),
        next_mldsa65_signing_raw=fill(MLDSA65_VERKEY_BYTES, 0x12),
    )


def _ed25519_verfer_qb64(raw32: bytes) -> str:
    if len(raw32) != 32:
        raise ValueError("Ed25519 public key must be 32 bytes")
    return Verfer(raw=raw32, code=MtrDex.Ed25519).qb64


def _blake3_digest_qb64(data: bytes) -> str:
    return Diger(ser=data, code=MtrDex.Blake3_256).qb64


def material_to_cesr(material: HybridKeyMaterial) -> dict[str, str]:
    ed_q = _ed25519_verfer_qb64(material.ed25519_signing_raw)
    mldsa_q = encode_large_fixed(
        CESR_MLDSA65_VERKEY, material.mldsa65_signing_raw, MLDSA65_VERKEY_BYTES
    )
    x25519_q = encode_large_fixed(
        CESR_X25519_PUBKEY, material.x25519_agreement_raw, X25519_PUBKEY_BYTES
    )
    mlkem_q = encode_large_fixed(
        CESR_MLKEM768_ENCAP, material.mlkem768_encap_raw, MLKEM768_ENCAP_BYTES
    )
    n_ed = _blake3_digest_qb64(material.next_ed25519_signing_raw)
    n_mldsa = _blake3_digest_qb64(material.next_mldsa65_signing_raw)
    return {
        "ed25519_signing": ed_q,
        "mldsa65_signing": mldsa_q,
        "x25519_agreement": x25519_q,
        "mlkem768_encap": mlkem_q,
        "next_ed25519_digest": n_ed,
        "next_mldsa65_digest": n_mldsa,
    }


def _hybrid_anchor(cesr: dict[str, str]) -> list[dict[str, Any]]:
    """IA-HYBRID-1 cipher-suite tag + KA keys in standard a anchor (OQ-2 D1)."""
    return [
        {
            "ia": CIPHER_SUITE_IA_HYBRID_1,
            "ka": [cesr["x25519_agreement"], cesr["mlkem768_encap"]],
        }
    ]


def build_hybrid_inception(material: HybridKeyMaterial) -> dict[str, Any]:
    """Return keri 1.1.x conformant hybrid icp via SerderKERI makify."""
    cesr = material_to_cesr(material)
    anchor = _hybrid_anchor(cesr)

    vs = versify(version=Version, kind=Serials.json, size=0)
    ked: dict[str, Any] = {
        "v": vs,
        "t": Ilks.icp,
        "d": "",
        "i": "",
        "s": "0",
        "kt": "1",
        "k": [cesr["ed25519_signing"], cesr["mldsa65_signing"]],
        "nt": "1",
        "n": [cesr["next_ed25519_digest"], cesr["next_mldsa65_digest"]],
        "bt": "0",
        "b": [],
        "c": [],
        "a": anchor,
    }

    serder = serdering.SerderKERI(sad=ked, makify=True)
    serder._verify()

    inception_event = dict(serder.sad)
    said = serder.said
    aid = serder.pre

    return {
        "aid": aid,
        "said": said,
        "inception_event": inception_event,
        "raw_bytes_b64": base64.b64encode(serder.raw).decode("ascii"),
        "cipher_suite": CIPHER_SUITE_IA_HYBRID_1,
        "cesr": cesr,
        "public_key": cesr["ed25519_signing"],
        "next_key_digest": cesr["next_ed25519_digest"],
    }


def generate_hybrid_key_material() -> HybridKeyMaterial:
    """Generate fresh classical + placeholder PQC material (PQC bytes filled for structure)."""
    import pysodium  # lazy: synthetic harness path does not need libsodium

    ed_seed = pysodium.randombytes(32)
    ed_pk, _ = pysodium.crypto_sign_seed_keypair(ed_seed)
    next_ed_seed = pysodium.randombytes(32)
    next_ed_pk, _ = pysodium.crypto_sign_seed_keypair(next_ed_seed)
    return HybridKeyMaterial(
        ed25519_signing_raw=ed_pk,
        mldsa65_signing_raw=pysodium.randombytes(MLDSA65_VERKEY_BYTES),
        x25519_agreement_raw=pysodium.randombytes(X25519_PUBKEY_BYTES),
        mlkem768_encap_raw=pysodium.randombytes(MLKEM768_ENCAP_BYTES),
        next_ed25519_signing_raw=next_ed_pk,
        next_mldsa65_signing_raw=pysodium.randombytes(MLDSA65_VERKEY_BYTES),
    )

File: identity-agent-core/m63/hybrid_inception.go
package m63

import (
	"crypto/ed25519"
	"encoding/base64"
	"fmt"
)

// HybridKeyMaterial holds raw key bytes for hybrid inception (synthetic or caller-supplied).
type HybridKeyMaterial struct {
	Ed25519SigningRaw     []byte
	MLDSA65SigningRaw     []byte
	X25519AgreementRaw    []byte
	MLKEM768EncapRaw      []byte
	NextEd25519SigningRaw []byte
	NextMLDSA65SigningRaw []byte
}

// SyntheticHybridKeyMaterial returns deterministic synthetic bytes for harness vectors.
func SyntheticHybridKeyMaterial(seed int) HybridKeyMaterial {
	fill := func(n, tag int) []byte {
		out := make([]byte, n)
		for i := range out {
			out[i] = byte((seed + tag + i) % 256)
		}
		return out
	}
	return HybridKeyMaterial{
		Ed25519SigningRaw:     fill(Ed25519PubkeyBytes, 0x01),
		MLDSA65SigningRaw:     fill(MLDSA65VerkeyBytes, 0x02),
		X25519AgreementRaw:    fill(X25519PubkeyBytes, 0x03),
		MLKEM768EncapRaw:      fill(MLKEM768EncapBytes, 0x04),
		NextEd25519SigningRaw: fill(Ed25519PubkeyBytes, 0x11),
		NextMLDSA65SigningRaw: fill(MLDSA65VerkeyBytes, 0x12),
	}
}

type cesrKeys struct {
	Ed25519Signing    string `json:"ed25519_signing"`
	MLDSA65Signing    string `json:"mldsa65_signing"`
	X25519Agreement   string `json:"x25519_agreement"`
	MLKEM768Encap     string `json:"mlkem768_encap"`
	NextEd25519Digest string `json:"next_ed25519_digest"`
	NextMLDSA65Digest string `json:"next_mldsa65_digest"`
}

// HybridInceptionResult mirrors the keripy driver response shape.
type HybridInceptionResult struct {
	AID            string                 `json:"aid"`
	SAID           string                 `json:"said"`
	InceptionEvent map[string]interface{} `json:"inception_event"`
	RawBytesB64    string                 `json:"raw_bytes_b64"`
	CipherSuite    string                 `json:"cipher_suite"`
	CESR           cesrKeys               `json:"cesr"`
	PublicKey      string                 `json:"public_key"`
	NextKeyDigest  string                 `json:"next_key_digest"`
}

func ed25519VerferQB64(raw32 []byte) (string, error) {
	if len(raw32) != Ed25519PubkeyBytes {
		return "", fmt.Errorf("Ed25519 public key must be 32 bytes")
	}
	_ = ed25519.PublicKey(raw32)
	return MatterFixedQB64("D", raw32)
}

func materialToCESR(m HybridKeyMaterial) (cesrKeys, error) {
	ed, err := ed25519VerferQB64(m.Ed25519SigningRaw)
	if err != nil {
		return cesrKeys{}, err
	}
	mldsa, err := EncodeLargeFixed(CESRMLDSA65Verkey, m.MLDSA65SigningRaw, MLDSA65VerkeyBytes)
	if err != nil {
		return cesrKeys{}, err
	}
	x, err := EncodeLargeFixed(CESRX25519Pubkey, m.X25519AgreementRaw, X25519PubkeyBytes)
	if err != nil {
		return cesrKeys{}, err
	}
	kem, err := EncodeLargeFixed(CESRMLKEM768Encap, m.MLKEM768EncapRaw, MLKEM768EncapBytes)
	if err != nil {
		return cesrKeys{}, err
	}
	nEd, err := Blake3QB64(m.NextEd25519SigningRaw)
	if err != nil {
		return cesrKeys{}, err
	}
	nMldsa, err := Blake3QB64(m.NextMLDSA65SigningRaw)
	if err != nil {
		return cesrKeys{}, err
	}
	return cesrKeys{
		Ed25519Signing:    ed,
		MLDSA65Signing:    mldsa,
		X25519Agreement:   x,
		MLKEM768Encap:     kem,
		NextEd25519Digest: nEd,
		NextMLDSA65Digest: nMldsa,
	}, nil
}

func hybridAnchor(cesr cesrKeys) []anchorSeal {
	return []anchorSeal{{
		Ia: CipherSuiteIAHybrid1,
		Ka: []string{cesr.X25519Agreement, cesr.MLKEM768Encap},
	}}
}

func wireFromCESR(cesr cesrKeys) icpWire {
	return icpWire{
		T:  "icp",
		S:  "0",
		Kt: "1",
		K:  []string{cesr.Ed25519Signing, cesr.MLDSA65Signing},
		Nt: "1",
		N:  []string{cesr.NextEd25519Digest, cesr.NextMLDSA65Digest},
		Bt: "0",
		B:  []interface{}{},
		C:  []interface{}{},
		A:  hybridAnchor(cesr),
	}
}

func wireToInceptionMap(w icpWire) map[string]interface{} {
	return map[string]interface{}{
		"v": w.V, "t": w.T, "d": w.D, "i": w.I, "s": w.S, "kt": w.Kt,
		"k": w.K, "nt": w.Nt, "n": w.N, "bt": w.Bt,
		"b": w.B, "c": w.C, "a": []map[string]interface{}{{
			"ia": w.A[0].Ia,
			"ka": w.A[0].Ka,
		}},
	}
}

// BuildHybridInception constructs keri 1.1.17 conformant hybrid icp (SerderKERI makify).
func BuildHybridInception(m HybridKeyMaterial) (*HybridInceptionResult, error) {
	cesr, err := materialToCESR(m)
	if err != nil {
		return nil, err
	}

	final, raw, err := makifyICPWire(wireFromCESR(cesr))
	if err != nil {
		return nil, err
	}

	return &HybridInceptionResult{
		AID:            final.I,
		SAID:           final.D,
		InceptionEvent: wireToInceptionMap(final),
		RawBytesB64:    base64.StdEncoding.EncodeToString(raw),
		CipherSuite:    CipherSuiteIAHybrid1,
		CESR:           cesr,
		PublicKey:      cesr.Ed25519Signing,
		NextKeyDigest:  cesr.NextEd25519Digest,
	}, nil
}